src/rgw/driver/posix/rgw_sal_posix.cc

```diff
@@ -231,24 +227,62 @@ void POSIXDriver::register_admin_apis(RGWRESTMgr* mgr)
   return next->register_admin_apis(mgr);
 }
 
+std::unique_ptr<Notification> POSIXDriver::get_notification(rgw::sal::Object* obj,
+			      rgw::sal::Object* src_obj, struct req_state* s,
+			      rgw::notify::EventType event_type, optional_yield y,
+			      const std::string* object_name)
+{
+  return next->get_notification(obj, src_obj, s, event_type, y, object_name);
+}
+
+std::unique_ptr<Notification> POSIXDriver::get_notification(const DoutPrefixProvider* dpp,
+                              rgw::sal::Object* obj, rgw::sal::Object* src_obj,
+                              rgw::notify::EventType event_type,
+                              rgw::sal::Bucket* _bucket,
+                              std::string& _user_id, std::string& _user_tenant,
+                              std::string& _req_id, optional_yield y)
+{
+  return next->get_notification(dpp, obj, src_obj, event_type, _bucket, _user_id, _user_tenant, _req_id, y);
+}
+
 int POSIXUser::list_buckets(const DoutPrefixProvider* dpp, const std::string& marker,
 			     const std::string& end_marker, uint64_t max,
 			     bool need_stats, BucketList &buckets, optional_yield y)
 {
   DIR* dir;
   struct dirent* entry;
+  int dfd;
   int ret;
 
   buckets.clear();
 
-  dir = fdopendir(driver->get_root_fd());
+  /* it's not sufficient to dup(root_fd), as as the new fd would share
```

Comment:
Did you try rewinddir()?

Reply:
yes of course, and it trivially "works," but it's not thread safe;  opening a new fd is independent

Reply:
(discussed this at in detail with Marcus, he and I both prefer this and think it's not expensive);  n.b., it also resolves the memory leak we think happens at fdopendir

Reply:
I didn't mess with list_bucket obviously, I figured I'd look at that when trying to merge the listing cache

Reply:
Okay, I misread and thought this was listing a bucket.